File: nicegui/elements/scene/objects/extrusion.js
import SceneLib from "nicegui-scene";
const {
  THREE,
} = SceneLib;

export default class Extrusion {
  create_geometry(outline, height) {
    const shape = new THREE.Shape();
    shape.autoClose = true;
    if (outline.length) {
      shape.moveTo(outline[0][0], outline[0][1]);
      outline.slice(1).forEach((p) => shape.lineTo(p[0], p[1]));
    }
    const settings = { depth: height, bevelEnabled: false };
    return new THREE.ExtrudeGeometry(shape, settings);
  }
}

File: nicegui/elements/scene/objects/extrusion.py
from ..scene_object3d import Object3D


class Extrusion(Object3D, component='extrusion.js'):
    def __init__(
        self,
        outline: list[list[float]],
        height: float,
        wireframe: bool = False,
    ) -> None:
        """Extrusion

        This element is based on Three.js' `ExtrudeGeometry <https://threejs.org/docs/index.html#api/en/geometries/ExtrudeGeometry>`_ object.
        It is used to create a 3D shape by extruding a 2D shape to a given height.

        :param outline: list of points defining the outline of the 2D shape
        :param height: height of the extrusion
        :param wireframe: whether to display the extrusion as a wireframe (default: `False`)
        """
        super().__init__(outline, height, wireframe)
